### ISSUE TITLE
[Announcement] Dropping Python 3.6 support

### ISSUE THREAD

**Issue by pukkandan** [yt-dlp/yt-dlp#3764]
~~#### This is not an announcement. We are simply discussing the possibility for now~~

Following https://github.com/yt-dlp/yt-dlp/issues/267, we have now gotten rid of almost all Python 2.7 compatibility code. Most of our remaining compat code is for working around Py3.6 limitations. Since Python 3.6 has already reached EOL, it is time to start considering dropping it.

Compared to the benefits of dropping 2.7, the benefits we gain here is admittedly much smaller:
1. Remove `compat/re`, `compat/asyncio`
2. `contextvars` (https://github.com/yt-dlp/yt-dlp/pull/2173#issuecomment-1017990110) and `dataclasses` (to replace `info_dict` etc) - While these are beneficial in theory, making actual use of these take quite a bit of work. So they are not immediately useful
3. #3668 needs 3.7 since `requests` module is about to drop support for 3.6

PS: This will not change our currently minimum supported Windows version (Vista)

~~Since the overall benefits is so small, we could alternatively wait for 3.7's EOL (2023-06-27) and then jump to 3.8 directly~~

IMPORTANT: https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1154051119

**Comment by Ashish0804** [comment]
Also since there are only 2 issues pinned now, maybe pin this aswell till u find another issue to pin. (i'm aware this issue is already mentioned in the discussion master pinned issue)

**Comment by BlohoJo** [comment]
Except, this means that it will no longer work on Windows 7, correct?

**Comment by pukkandan** [comment]
> Except, this means that it will no longer work on Windows 7, correct?

No, windows 7 supports upto 3.8

**Comment by BlohoJo** [comment]
Oh right, sorry about that, for some reason I thought 3.6 was the highest on Windows 7. Don't know why I thought that! :p

**Comment by coletdjnz** [comment]
> Since the overall benefits is so small, we could alternatively wait for 3.7's EOL (2023-06-27) and then jump to 3.8 directly

I'm in favor of dropping it now, and when 3.7 reaches EOL, we drop that then too, and so on (with some advance warning). We should not be expected to support EOL python versions, **especially** as yt-dlp is an internet-based application and EOL python versions do not receive security updates. 

I think it's better to incrementally drop python versions as they become EOL at a minimum, likewise to incrementally bring new features and improvements into the codebase. We do not want to fall into the youtube-dl trap of unnecessary legacy support.

> lets drop py3.6 and 3.7 right now only and jump straight to 3.8

as much as I'd like too, since it's not EOL, we can still expect it to be used for the time-being.

**Comment by gamer191** [comment]
> Oh right, sorry about that, for some reason I thought 3.6 was the highest on Windows 7. Don't know why I thought that! :p

See https://github.com/yt-dlp/yt-dlp/issues/390 for the highest supported versions on each OS
> I'm in favor of dropping it now, and when 3.7 reaches EOL, we drop that then too, and so on (with some advance warning). We should not be expected to support EOL python versions, **especially** as yt-dlp is an internet-based application and EOL python versions do not receive security updates.

In my opinion there should always be a discussion about whether to drop support for old versions of Python, rather than just an announcement.

**Comment by gamer191** [comment]
On a different note, I don't believe python 3.6 support should be dropped, until https://github.com/yt-dlp/yt-dlp/issues/2701 is fixed, because we can't expect users who want to work around that bug to use python 3.10+

**Comment by coletdjnz** [comment]
> On a different note, I don't believe python 3.6 support should be dropped, until https://github.com/yt-dlp/yt-dlp/issues/2701 is fixed, because we can't expect users who want to work around that bug to use python 3.10+

kinda same thing goes the other way? can't expect users who want to work around that bug to use python 3.6 (as they likely do not have 3.6 installed and we should not be suggesting them to do so).

I don't think an extractor/site-specific issue should hold back the entire project. Especially when there is a proposed fix (see the PR).

That said we can probably get that site working before we would end up dropping 3.6 anyways if it is a problem

**Comment by gamer191** [comment]
> bruh u arguing like dlp is some utility subsidised by UN

I'm not arguing. Pukkandan created this discussion, for the purpose of people sharing their opinions on dropping python 3.6 support. I was just sharing my opinion on the topic (and, for the record, I use Python 3.10, so this doesn't even affect me).

> last time i remember some dumbass tried to make issues and PRs to reintroduce py2 months after it was dropped

Well they clearly cared a lot if they went to the effort of making a PR for it! Or was the PR just to update the readme?🤣
> dlp shouldn't care what most people are using or what eol is. Just support officially support whichever version the owner is running imo.

I don't disagree with that. But it's up to the @pukkandan to weigh up whether to make yt-dlp easier to contribute for developers or easier to use for users.

> same way till 3.7 is dropped, dlp can show this https://www.indeed.com/career-advice/finding-a-job/better-job as warning to people using ancient machines and eol versions.

For clarity, I don't oppose removing python 3.7 support (I actually haven't thought about it much). I just think that there should be a discussion first, but that's up to @pukkandan (see my above point).



> Maybe it is time to consider dropping Vista support

@d1nuc0m yt-dlp will most likely work on Vista until Python 3.7 support is dropped, when it will break.

**Comment by fstirlitz** [comment]
> Some devices aren't compatible with 3.10, but are compatible with 3.6.

As in, they are incapable of having their software upgraded at all, or is it just a case of users too lazy to switch operating systems? I happen to run Debian unstable (the ‘rolling release’ version) on a 32-bit laptop from 2003 or so. It runs 3.10 just fine. I don’t expect it to play 1080p video smoothly, but for streaming music in the background while coding it’s good enough.

Moving to 3.7 may be warranted because of `contextvars` (and to a lesser extent `dataclasses`; that one is shimmable relatively easily), but I don’t think the [features of 3.8](https://docs.python.org/3/whatsnew/3.8.html) are nearly compelling enough to force a switch:

- The walrus: dispensable, you can always go back to writing assignment as a separate statement
- Positional-only parameters: nice, but can be approximated with a decorator
  ```python
  def positional_only(*params, invert=False):
      import inspect, functools
      def decorate(func):
          sig = inspect.signature(func)
          assert frozenset(params).issubset(sig.parameters.keys())
          assert all(sig.parameters[param].kind in (
              inspect.Parameter.POSITIONAL_ONLY,
              inspect.Parameter.POSITIONAL_OR_KEYWORD,
          ) for param in params)
          newsig = sig.replace(parameters=(
              param.replace(kind=inspect.Parameter.POSITIONAL_ONLY)
              if (param.name in params) != invert else param
              for param in sig.parameters.values()
          ))
          @functools.wraps(func)
          def wrapper(*args, **kwargs):
              bargs = newsig.bind(*args, **kwargs)
              return func(*bargs.args, **bargs.kwargs)
          wrapper.__signature__ = newsig
          return wrapper
      return decorate  
  ```
  Introducing positional-only parameters to existing calls might necessitate a deprecation cycle first, which would require going through a decorator anyway. So the language feature can wait, I think.
- Postfix `=` in f-strings: dispensable debugging-only feature, useful primarily in WIP code.
- Everything else: irrelevant

3.9 has even fewer compelling [features](https://docs.python.org/3/whatsnew/3.9.html) than that (pretty much none). 3.10 at least adds pattern matching: that one might be useful, but other [prominent features](https://docs.python.org/3/whatsnew/3.10.html) are either entirely irrelevant or development-only improvements. I don’t think pattern matching alone is compelling enough.

3.11 may be the first release worth breaking backwards compatibility for, since it is going to add exception groups and `TaskGroup`, which may simplify asynchronous programming a lot. `TaskGroup` alone may be shimmable (with some effort), but exception groups will be hard to simulate. However, it’s going to take a while until 3.11 becomes widely available.

Also note that requiring any version later than 3.9 amounts to dropping support for PyPy, because this is the latest Python version it supports. And even that is not always available.

**Comment by pukkandan** [comment]
> Some devices aren't compatible with 3.10, but are compatible with 3.6.

This is perhaps the most important question here. Is there actually any device that supports py3.6, but not 3.7/8? Or at-least, is there still any widely used UNIX distro that comes with only 3.6? If there is not, even minimal improvements are enough reason. After all, we can weigh the cost/benefit only if there is any cost to begin with.

**Comment by d1nuc0m** [comment]
> yt-dlp will most likely work on Vista until Python 3.7 support is dropped, when it will break.

Yes, what I wanted to say is "breaking Vista compatibility shouldn't be too much a problem"

> Youtube-dl is practically abandoned

They "changed management", we'll see if they'll be able to manage all the backlog

**Comment by gamer191** [comment]
> Is there actually any device that supports py3.6, but not 3.7/8?

I didn't say that. I said that there are devices that support 3.6, but not 3.10. This is relevant, because https://github.com/yt-dlp/yt-dlp/issues/2701 only affects python 3.7-9

So the cost of dropping python 3.6 support is breaking https://github.com/yt-dlp/yt-dlp/issues/2701 on Windows Vista and Windows 7 devices (I don't know which non-Windows devices are affected). 

> > Youtube-dl is practically abandoned
> 
> They "changed management", we'll see if they'll be able to manage all the backlog

Yeah, back in January

**Comment by pukkandan** [comment]
> This is relevant, because #2701 only affects python 3.7-9
> 
> So the cost of dropping python 3.6 support is breaking #2701 on Windows Vista and Windows 7 devices (I don't know which non-Windows devices are affected).

https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1129771692

**Comment by Evernow** [comment]
> So the cost of dropping python 3.6 support is breaking #2701 on Windows Vista and Windows 7 devices (I don't know which non-Windows devices are affected).

Acceptable if you ask me. Those users (especially Vista) shouldn't even be facing the internet.

**Comment by sulyi** [comment]
There might not be much demand for running yt-dlp on embedded systems. And I'm not even sure if they will be effected by this.
But please, don't forget to consider them.

**Comment by d1nuc0m** [comment]
> There might not be much demand for running yt-dlp on embedded systems. And I'm not even sure if they will be effected by this.

Embedded systems should run with modern Linux distros (as Windows CE is going to be EOL soon), so this shouldn't be a problem

**Comment by wanghaisheng** [comment]
+1

**Comment by Vangelis66** [comment]
> PS: This will not change our currently minimum supported Windows version (Vista)
> 
> Since the overall benefits is so small, we could alternatively wait for 3.7's EOL (2023-06-27)

I sincerely hope this original plan is stuck to... I can't speak about the pros and cons of **py3.7** on non-Windows OSes, but indeed py3.7 is EoS for **Vista SP2**...
As noted by others, **py3.7** is not EoL'ed yet by the **PSF**, only compiled binaries have stopped being released by them, however **third party binaries** of py3.7 (latest being [v3.7.13](https://github.com/adang1345/PythonWindows/tree/master/3.7.13)) are available to Windows users, **full with the latest security patches** issued by the **PSF**...

As for the Vista-haters, let me inform **the rest** of you (as the opinion of the haters simply can't be changed) that the OS is enjoying popularity among the retro-computing communities, especially after the creation of the **VistaExtendedKernel** project, and those users that have stayed on the OS on somewhat older but still worthy H/W **have their systems carefully and fully patched** with vendor updates from Vista's Server counterpart (WS2008), that'll be available for at least until **Jan 2023** (same time as **Win7** ESU updates end); you can learn more, if you need, in dedicated places like [the MDL forums](https://forums.mydigitallife.net/threads/updated-how-to-install-and-update-windows-vista-sp2-64-bit-beyond-its-eol-date.83461/) and elsewhere (**MSFN**, etc.). Using such systems to access the web should be as safe (or, perhaps, more) as using the still plentiful **Win7 SP1** systems out there, most of which have stayed on a MS patch level of Jan 2021 (FTR, **Security Suites** and **currently maintained web browsers,** full with vendor patches, are still available for the OS - don't take the MS FUD verbatim: you won't be made "part of the botnet" the next day MS ends Extended Support for your OS) ... 

So, please, do not prematurely end `yt-dlp`'s support for **py3.7** - **13** additional months of support by the **PSF** is still a lot... And from what I read previously in this "discussion", the cost of maintaining such support programmatically should be **minimal** (disclaimer: I'm not a Python coder myself, this is my overall understanding of things). 

Many thanks to the whole `yt-dlp` team for all their unwavering efforts 👍 .

**Comment by sulyi** [comment]
> Embedded systems should run with modern Linux distros

No. In fact dropping 2.6 - 3.6 probably already effected a lot, so it's not a big deal one minor version number more.

Just saying design decisions like like monolithic distribution, and support of EOL interpreters was originally made around supporting embedded systems (I mean mainly routers). But I'm fine with not carrying that baggage.

To stop fixing EOL version specific issues makes sense, else why drop others and not this one.
To try not to break them if avoidable is, however sympathetic effort ultimately, nonsense.

**Comment by dirkf** [comment]
>EDIT: older OSes would still be supported by yt-dl, that aims to use Python 2.6, 2.7, 3.2+

However the current symbiosis might be damaged by wider adoption of novel Python syntax in yt-dlp. Obviously there will be some cases (async, eg) where yt-dlp code may need advanced features, but unnecessary adoption of novel syntax would hinder cross-fertilisation between yt-dlp and yt-dl.

Generally, novel programming language syntax is destabilising. Like internet protocols, the syntax is a hard interface where the programmer should be strict in what is sent and the language translator should be liberal in what it receives. Linters exist to enforce the strict side. There are plenty of examples of unnecessary pain inflicted by syntax changes to JavaScript that should have been functions (compare the ES2020 `?.` optional chaining operator with  `try_get()`/`traverse_obj()` used here).

>... Is there actually any device that supports py3.6, but not 3.7/8? Or at-least, is there still any widely used UNIX distro that comes with only 3.6?

The Ubuntu repository for 16.04ESM, supported to 2026, offers Py3.5 although 3.9 and 3.10 are available through a reliable PPA. But then the system actually depends on 2.7. Probably this doesn't count as widely used, though, even if its pedigree is good.

**Comment by MrRawes** [comment]
> > EDIT: older OSes would still be supported by yt-dl, that aims to use Python 2.6, 2.7, 3.2+
> 
> However the current symbiosis might be damaged by wider adoption of novel Python syntax in yt-dlp. Obviously there will be some cases (async, eg) where yt-dlp code may need advanced features, but unnecessary adoption of novel syntax would hinder cross-fertilisation between yt-dlp and yt-dl.
> 
> Generally, novel programming language syntax is destabilising. Like internet protocols, the syntax is a hard interface where the programmer should be strict in what is sent and the language translator should be liberal in what it receives. Linters exist to enforce the strict side. There are plenty of examples of unnecessary pain inflicted by syntax changes to JavaScript that should have been functions (compare the ES2020 `?.` optional chaining operator with `try_get()`/`traverse_obj()` used here).
> 
> > ... Is there actually any device that supports py3.6, but not 3.7/8? Or at-least, is there still any widely used UNIX distro that comes with only 3.6?
> 
> The Ubuntu repository for 16.04ESM, supported to 2026, offers Py3.5 although 3.9 and 3.10 are available through a reliable PPA. But then the system actually depends on 2.7. Probably this doesn't count as widely used, though, even if its pedigree is good.

whats wrong with adding the ppa or simply updating the OS

**Comment by gamer191** [comment]
> The Ubuntu repository for 16.04ESM, supported to 2026, offers Py3.5 although 3.9 and 3.10 are available through a reliable PPA. But then the system actually depends on 2.7. Probably this doesn't count as widely used, though, even if its pedigree is good.

How does any of that relate to whether Py3.6 should be dropped?

**Comment by dirkf** [comment]
It responds to the quoted question. A major distro provider is sticking at 3.5 and not **even** 3.6.

**Comment by gamer191** [comment]
> However the current symbiosis might be damaged by wider adoption of novel Python syntax in yt-dlp. Obviously there will be some cases (async, eg) where yt-dlp code may need advanced features, but unnecessary adoption of novel syntax would hinder cross-fertilisation between yt-dlp and yt-dl.

I somewhat agree with you, [but it's a bit late for this discussion](https://github.com/yt-dlp/yt-dlp/commit/86e5f3ed2e6e71eb81ea4c9e26288f16119ffd0c). Admittedly, I was the person who initially suggested running pyupgrade on yt-dlp, but I later regretted it

**Comment by dirkf** [comment]
As that's one commit, it might be possible to roll it back on a yt-dlp checkout before back-porting ... But those are all trivial changes that need to be caught. Things like `nonlocal`, though a useful construct, need more elaborate work-arounds.

**Comment by fstirlitz** [comment]
> As for the Vista-haters, let me inform the rest of you (as the opinion of the haters simply can't be changed) that the OS is enjoying popularity among the retro-computing communities, especially after the creation of the VistaExtendedKernel project

I’m all for retrocomputing, but when you take up such a hobby, it’s generally understood that when things don’t work, the responsibility for fixing them rests primarily on you. You chose an unusual configuration and you knew it; it’s your job to support it instead of making demands of other people. And the sort of person who maintains their own patched kernel for a proprietary OS should have no trouble getting newer Python versions to run, whether the original binaries or a custom build. (I suspect it wouldn’t even be that hard.)

Personally, I think switching the OS would be less effort in the long run and a more rational option, especially for PC-based devices (embedded stuff is another matter) where the process is well-understood, relatively easy and low-risk (as long as repartitioning succeeds). No need to lazily label people as ‘haters’; not that I particularly care for Windows, but it’s your post that seems primarily emotion-driven.

**Comment by WassimAttar** [comment]
Anyone knows who are the yt-dlp users ?
Companies ?
Individuals ?

Anyone knows on which devices yt-dlp is running on ?
Embeded ?
Routers ?
NASes ?
Raspbery Pies ?
PCs ?

Anyone knows on which OSes yt-dlp us running on ?
x86 linux ?
arm linux ?
windows ?
mac Os ?

Collecting data may give some insigths for decision making.

**Comment by Ashish0804** [comment]
> Collecting data may give some insigths for decision making.

dlp is not a paid product that warrants doing a customer survey imo. This post here is already more than that is required imo.

> As for the Vista-haters, let me inform the rest of you (as the opinion of the haters simply can't be changed) that the OS is enjoying popularity among the retro-computing communities

> especially after the creation of the VistaExtendedKernel project

I'm sure the community is capable enough to port py3.8 to vista aswell then.
So that means maintainers of dlp should be good to drop py3.6 and 3.7 support

Glad we could get vista-bros behind this.

**Comment by nicolaasjan** [comment]
Will yt-dlp then stop working on Ubuntu 18.04 LTS and derivates (ships with Python 3.6.8)?

Ubuntu 18.04 is supported till April 2023 (standard support)
https://wiki.ubuntu.com/Releases

**Comment by Evernow** [comment]
> Ubuntu 18.04 LTS and derivates (ships with Python 3.6.8)?

You can get even 3.10 on 18.04 via PPAs or just building from source.

**Comment by Evernow** [comment]
> So that means maintainers of dlp should be good to drop py3.6 and 3.7 support

I also support dropping 3.7 support now, so that this same discussion doesn't have to happen again next year. Users who depend on such old versions should expect not to be able to use the newest version of software. 

Rising requirement to 3.8 would mean this conversation wouldn't have to happen for more than 2 years.

**Comment by dirkf** [comment]
But [fstirlitz](https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1131307029) convincing analysis* suggests that only 3.7 or 3.11 would offer significant advantages, bearing in mind that this is a question of what language version contributors write to, and not what implementation version is recommended for running the program.
<small>
*except for this:

>Positional-only parameters: nice, ...

PEP-570 seemed to me somewhat spurious but then I may be late to the Python party. Having been entertained by TCL, I would have thought a more logical and useful change would have been to allow any parameter to be called as a keyword argument, provided that all subsequent parameters in the call are also called in that way. Anyhow, +1 for the decorator.
</small>

**Comment by pukkandan** [comment]
> I would have thought a more logical and useful change would have been to allow any parameter to be called as a keyword argument, provided that all subsequent parameters in the call are also called in that way

The (only?) advantage of positional-only arguments is that APIs do not have to expose the argument name

**Comment by pukkandan** [comment]
> Will yt-dlp then stop working on Ubuntu 18.04 LTS and derivates (ships with Python 3.6.8)?
> 
> Ubuntu 18.04 is supported until April 2023 (standard support) [wiki.ubuntu.com/Releases](https://wiki.ubuntu.com/Releases)



We could release Linux standalone builds so users don't have to worry about installed python version: 
Eg: https://github.com/pukkandan/yt-dlp-dev/releases/tag/2022.05.22.113344
Code: https://github.com/pukkandan/yt-dlp-dev/tree/features/unix-build

**Comment by nicolaasjan** [comment]
> We could release Linux standalone builds. With this, all that matters is whether the OS supports the Python version, not whether the Python is pre-installed.

That worked. :)
I didn't even know you had this `yt-dlp-dev`. Where is that hidden?

I only had to give my GitHub account password to run it (?)
Then did `./yt-dlp_linux -Uv` and it updated to the next version.

```
./yt-dlp_linux -Uv
/media/ext3-data/Downloads/ytdlp_plugins/postprocessor/__init__.py
/media/ext3-data/Downloads/ytdlp_plugins/extractor/__init__.py
('linux_exe', '/media/ext3-data/Downloads/yt-dlp_linux')
[debug] Command-line config: ['--ignore-config', '-Uv']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
('linux_exe', '/media/ext3-data/Downloads/yt-dlp_linux')
[debug] yt-dlp version 2022.05.22.113353 [d42ea93] (linux_exe)
[debug] Python version 3.10.4 (CPython 64bit) - Linux-5.4.0-110-generic-x86_64-with-glibc2.31
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: ffmpeg -bsfs
[debug] exe versions: ffmpeg N-106448-gc2d1597a8a-Nico-20220328 (fdk,setts), ffprobe N-106448-gc2d1597a8a-Nico-20220328, phantomjs 2.1.1
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2022.05.18.1, mutagen-1.45.1, secretstorage-3.3.2, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
Latest version: 2022.05.22.113353, Current version: 2022.05.22.113353
yt-dlp is up to date (2022.05.22.113353)

```

**Comment by pukkandan** [comment]
> I didn't even know you had this yt-dlp-dev. Where is that hidden?

It's my personal fork. Nothing others need to know about generally

> I only had to give my GitHub account password to run it (?)

wdym?

**Comment by nicolaasjan** [comment]
> wdym?

Well, as I said, the first time I ran it (after making it executable), it prompted for my GH account password...
After it updated, it ran without.

**Comment by pukkandan** [comment]
> it prompted for my GH account password

"It" = `yt-dlp_macos` file?

There is no reason yt-dlp should ask for your GH password... Are you sure you didn't run something else?

**Comment by Lesmiscore** [comment]
I think you mispelt URL for git clone?
That would ask for password.

**Comment by nicolaasjan** [comment]
> > it prompted for my GH account password
> 
> "It" = `yt-dlp_macos` file?
> 
> There is no reason yt-dlp should ask for your GH password... Are you sure you didn't run something else?

No, it was the `yt-dlp_linux` file.

**Comment by nicolaasjan** [comment]
> I think you mispelt URL for git clone? That would ask for password.

I'm pretty sure I just downloaded the `yt-dlp_linux` file by clicking on the link under [Releases](https://github.com/pukkandan/yt-dlp-dev/releases).

But as I said, it didn't happen doing it a second time.
So this remains a mystery.

**Comment by LTIOfficial** [comment]
For the record, I am running it on a CENTOS 7 install (based on RedHat Enterprise), although the following applies equally to numerous other 'enterprise-grade' operating systems that include full support for most of their entire 10-year lifespan, followed by backporting of security/bug fixes over their last 3-4 years, depending on distro.  In CENTOS' case, the EOL (and therefore support for Python 3.6) falls on June 30, 2024.  Unfortunately, I am not aware of the EOL dates for other enterprise distros to keep Python 3.6 bug and vulnerability-free.

**Comment by coletdjnz** [comment]
> For the record, I am running it on a CENTOS 7 server install (based on RedHat Enterprise), although the following applies equally to numerous other 'enterprise-grade' operating systems that include full support for most of their entire 10-year lifespan, followed by backporting of security/bug fixes over their last 3-4 years, depending on distro.  In CENTOS' case, the EOL (and therefore support for Python 3.6) falls on June 30, 2024.  Unfortunately, I am not aware of the EOL dates for other enterprise distros to keep Python 3.6 bug and vulnerability-free.
> 
> So it is very much my hope that support for 3.6 is not dropped for at least the remaining life of my server.

I think this is a case where having a Linux binary with a Python interpreter bundled would be useful. Currently we do have a Linux binary but it doesn't come with python.

IMO we should not have to support 3.6 (and 3.7 for that matter) till 2024.

**Comment by Evernow** [comment]
@LTIOfficial 

> the EOL (and therefore support for Python 3.6) falls on June 30, 2024. 

This is not correct, as per RedHat's official documentation: https://access.redhat.com/support/policy/updates/rhscl-rhel7

Python 3.6 is end of life on RHEL 7 as of Oct 2020, only Python 3.8 is supported until 2024.

**Comment by pukkandan** [comment]
@LTIOfficial Wouldn't the standalone builds in https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1133877078 be satisfactory for your case?

**Comment by MrRawes** [comment]
a little offtopic but are there any plans to replace phantomjs with something else [since phantomjs is dead](https://github.com/ariya/phantomjs/issues/15344)

**Comment by pukkandan** [comment]
Since dropping **3.7** would (1) break Vista support (however small the userbase is), (2) We have no immediate use for 3.8 features, and (3) 3.7 is not EOL yet, we shouldn't drop support for it yet. We can do so when at least one of these points becomes no longer true.

For dropping 3.6, the main concern as I understand is that Ubuntu 16, CentOS7 (and possibly some other distros no one has yet named) comes with only 3.6. However, they do support newer versions of Python. So we [can just release standalone binaries](https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1133877078) to address these concerns.

Everyone using such OSes, please test the [POC standalone binary](https://github.com/pukkandan/yt-dlp-dev/releases/tag/2022.05.22.113344) and let me know in case it does **not** work. **In the absence of any such reports, we will proceed with dropping 3.6**

In order to bring this matter to the attention of more users, I will add a warning in the next release (which should be this week) when yt-dlp detects it is running on 3.6. The release after that will be the last one with 3.6 support, and updates (`-U`) will be blocked for any 3.6 users. (This is similar process we used for 2.7 deprecation)

PS: By this time, I expect to also have an implementation of #1839. In case I manage to implement it before dropping 3.6, the process will be tweaked so that 3.6 is dropped in just over a month from now

**Comment by fstirlitz** [comment]
Dataclasses are in 3.7 already. Though they are relatively easily backported to 3.6; it’s fewer than 100 lines of Python to accomplish. We might add a backport anyway if 3.10+ features like the `kw_only=` flag prove useful.

**Comment by gamer191** [comment]
> updates (`-U`) will be blocked for any 3.6 users

Why not just update 3.6 users to the standalone binaries?

**Comment by gamer191** [comment]
> For dropping 3.6, the main concern as I understand is that Ubuntu 16

based on looking through this conversation (and running `ls -ls /usr/bin/python*` on both Ubuntu 16.04 and 18.04), I'm assuming that that is a typo for Ubuntu 18.04, rather than 16.04.

I tested the POC binary on Ubuntu 18.04 and it worked :) but I did notice that it prints:
```
/home/ubuntu/Downloads/ytdlp_plugins/postprocessor/_init_.py
/home/ubuntu/Downloads/ytdlp_plugins/extractor/_init_.py
('linux_exe', '/home/ubuntu/Downloads/yt-dlp_linux')
```
at the start of every output, when it's run from `/home/ubuntu/Downloads`, despite no directory existing called `/home/ubuntu/Downloads/ytdlp_plugins`.

So, for example, `--version` prints

```
/home/ubuntu/Downloads/ytdlp_plugins/postprocessor/_init_.py
/home/ubuntu/Downloads/ytdlp_plugins/extractor/_init_.py
('linux_exe', '/home/ubuntu/Downloads/yt-dlp_linux')
2022.05.22.113344
``` 
On a different note, for those pip users who have never run an executable file through the terminal, it can be run using `./yt-dlp_linux`, after making sure to enable `allow executing file as program` (or something like that) in the files properties

**Comment by Evernow** [comment]
> That said, someone could probably self-compile python for 32-bit devices, so it's arguably no different to when Python gets a source-only release

Python 3.7 does not release binaries for **any** architecture or operating system anymore: https://peps.python.org/pep-0537/#schedule-security-binary-release , so given that as you mentioned there is no real difference between that and a source-only release, I believe it's fine. 

The same applies for 3.6 and [3.8](https://peps.python.org/pep-0569/#source-only-security-fix-releases).

I believe there is a precedent anyways, Microsoft provides 32-bit Python builds for macOS in [actions](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-12-Readme.md#python)

**Comment by gamer191** [comment]
@evernow I just edited my message 5 seconds after you posted that 🤣

It's not the same as a source-only release, because ource-only releases are normally just for security updates, and 3.7.7 is a bugfix update

**Comment by gamer191** [comment]
> Microsoft provides 32-bit Python builds for macOS in [actions](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-12-Readme.md#python)

Um, how do you know those are 32-bit? The ones which are available in [actions/setup-python](https://github.com/actions/python-versions/blob/main/versions-manifest.json) are all x64

**Comment by Lesmiscore** [comment]
> > Why not just update 3.6 users to the standalone binaries, if they are using Linux?
> 
> I don't think it's right to self-update from one variant to another. Users wouldn't expect that behavior


Agree. Who wants to get x86_64 binary on aarch64 machine!

**Comment by gamer191** [comment]
> I don't think it's right to self-update from one variant to another. Users wouldn't expect that behavior

Fair enough, perhaps a warning should be shown instead when the user attempts to run `yt-dlp -U`.

**Comment by gamer191** [comment]
> Agree. Who wants to get x86_64 binary on aarch64 machine!

I don't know enough about processors, so I don't know if that was a joke or serious, but if it's serious, then it brings up another possible issue with dropping 3.6 support, because someone might be running ubuntu 16.04 on an aarch64 (or a different instruction set) machine

**Comment by candrapersada** [comment]
will there be a future yt-dlp_win7.exe with Python 3.8 like yt-dlp_linux?

**Comment by gamer191** [comment]
> Will support for Python 3.8 on Windows 7 be discontinued in the future?

Probably not until at least October 2024. Is there a reason why you need to know?

**Comment by pukkandan** [comment]
Eventually, yes obv. I don't plan on supporting EOL software for perpetuity. But not for now

**Comment by gamer191** [comment]
> just wondering what will happen to the fate of Windows 7 users if support for Python 3.8 is discontinued as well.

They won't receive future updates, and yt-dlp will eventually break for them. If it's still around, they will be able to use [youtube-dl](https://github.com/ytdl-org/youtube-dl), but at the moment youtube-dl is light years behind yt-dlp, so don't get your hopes up.

There **might** also be workarounds for them (eg: virtualbox), but I'm not going to get into that now

It's also quite possible that python 3.8 support won't ever be removed, given part of the reason why python 3.6 support was discontinued was because practically no one needed it. That will be different for both python 3.7 and 3.8, so we'll have to wait and see what the maintainers at the time decide to do. I wouldn't be surprised if python 3.8 was supported well into 2025 (or even 2026). After all, python 3.6 could have been deprecated last December, but it wasn't even discussed until May this year.

It's also quite possible that there will be a very useful feature which will justify removing python 3.8 support before October 2024. After all, the Python software foundation has already stopped supporting it beyond security updates.

And what if yt-dlp has such a good generic extractor by then that youtube support never breaks (this is wishful thinking, I know)

Basically, what I'm trying to say is that lots and lots of things could happen. But I don't see any need to talk about it now. We can save that for when Pukkandan opens a discussion about dropping Python 3.8 support (although I, personally, probably won't be following yt-dlp's issue tracker by then).



> Eventually, yes obv. I don't plan on supporting EOL software for perpetuity. But not for now

I can't be bothered to re-write paragraph 3 of my above message. I think you get the general picture

**Comment by anunasica** [comment]
Hi, what should I do to be able in establishing if I have python 3.6 or the upcoming (perhaps not supported here!) 3.7 ?
I guess I have to type a code into cmd.exe ? thanks a lot.

**Comment by pukkandan** [comment]
Run `yt-dlp -v` and see the line starting with `[debug] Python version`
```
❯ yt-dlp -v
[debug] Command-line config: ['--ignore-config', '-v']
[debug] Encodings: locale cp65001, fs utf-8, pref cp65001, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.06.22.1 [a86e01e74] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: 28163422a
[debug] Python version 3.10.4 (CPython 64bit) - Windows-10-10.0.22000-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg N-106550-g072101bd52-20220410 (fdk,setts), ffprobe N-106624-g391ce570c8-20220415, phantomjs 2.1.1
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.1
[debug] Proxy map: {}

Usage: yt-dlp [OPTIONS] URL [URL...]

yt-dlp: error: You must provide at least one URL.
Type yt-dlp --help to see a list of all options.
```

**Comment by gamer191** [comment]
> upcoming (perhaps not supported here!) 3.7

3.7 is not upcoming, it's from 2018. It's the lowest version of Python that is supported, with 3.10 (technically 3.11, but that hasn't come out yet) being the highest

**Comment by pukkandan** [comment]
> technically 3.11, but that hasn't come out yet) being the highest

Technically, [3.12 is the latest](https://docs.python.org/3.12). 3.11 is in beta and 12 in alpha

**Comment by anunasica** [comment]
Thanks for all your replies, very helpful.
I do not trust to my eyes but I read "Python version 3.7.9 (CPython 32bit)". So I am fine for future YT-DLP releases! very good!!!! then I updated 05-18 to the latest 06-22 without any warning! wow :)))

**Comment by gamer191** [comment]
@anunasica What OS (Windows, Mac, Linux etc) are you running?

**Comment by thinkpad4** [comment]
> Will support for Python 3.8 on Windows 7 be discontinued in the future?
> and will there be a way to still be able to use yt-dlp in Windows 7?

Python 3.8 is supported until October 2024

**Comment by gnanet** [comment]
For Debian 9 (stretch) users: there is a way to have Python 3.7.13 (released 2022-03-14) installed, but it requires the compilation from source. **Tested right now, and its working**

## based on https://linuxize.com/post/how-to-install-python-3-7-on-debian-9/

### Start by installing the packages necessary to build Python source:

```
sudo apt update
sudo apt install build-essential zlib1g-dev libncurses5-dev libgdbm-dev libnss3-dev libssl-dev libreadline-dev libffi-dev curl libbz2-dev
```

### Download the latest release’s source code from the Python download page using the following curl command :

```
curl -O https://www.python.org/ftp/python/3.7.13/Python-3.7.13.tar.xz
```
### At the time of writing this article, the latest release is 3.7.13. (file release date 2022-03-16)

### When download is complete, extract the tarball :

```
tar -xf Python-3.7.13.tar.xz
```

### Navigate to the Python source directory and run the configure script that will perform a number of checks to make sure all of the dependencies on your system are present:

```
cd Python-3.7.13
./configure --enable-optimizations
```

### The --enable-optimizations option will optimize the Python binary by running multiple tests which will make the build process slower.

### Run make to start the build process:

```
make -j 8
```

### For faster build time, modify the -j flag according to your processor. If you do not know the number of cores in your processor you can find it by typing nproc. My system has 8 cores, so I am using the -j8 flag.

### Once the build is done install the Python binaries by running the following command as a user with sudo access :

```
sudo make altinstall
```

### Do not use the standard make install as it will overwrite the default system python3 binary.

### At this point, Python 3.7 is installed on your Debian system and ready to be used. You can verify it by typing:

```
python3.7 --version
```

**Comment by anunasica** [comment]
@gamer191:
OS is Vista32SP2.

**Comment by gamer191** [comment]
> For Debian 9 (stretch) users: there is a way to have Python 3.7.13 (released 2022-03-14) installed, but it requires the compilation from source

Can't Debian 9 users use `yt-dlp_linux` (which doesn't require Python to be installed)?

**Comment by gnanet** [comment]
> > For Debian 9 (stretch) users: there is a way to have Python 3.7.13 (released 2022-03-14) installed, but it requires the compilation from source
> 
> @gnanet Can't Debian 9 users use `yt-dlp_linux` (which doesn't require Python to be installed)?

@gamer191  Unfortunately no. It's because the default **libc** in Debian 9 is lower than the required 2.50

The whole stuff, about compiling 3.7 myself, resulted from my own try to get yt-dlp on my stretch.

**Comment by pukkandan** [comment]
The build has to be created against the minimum version we want to support. Since the [oldest Linux GHA supports is Ubuntu 18.04](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) which [is based on Debian 10](https://askubuntu.com/a/445496), it won't be possible for us to create standalone builds for Debian 9. May I ask why you don't upgrade?, especially considering even it's [LTS support expires in a few days](https://www.debian.org/releases/stretch/)

**Comment by gnanet** [comment]
I have no problems with the build standards, i just realized, and documented: it is easier to have py3.7 on Debian 9 than libc >= 2.50.

Ad system upgrade:
If the system is loaded with lots of stuff and its kind of VPS where a system and kernel upgrade is not supported, and the provider suggest to use a new VPS, and be satisfied with not able to keep your IP address, then I go rather with simply compile python 3.7.13, and happily use yt-dlp for a bit more time.

But i agree: for those who are able to have their system upgraded, it is advised to go that way.

**Comment by nicolaasjan** [comment]
> do i just download it to download folder and use it? or do i have to do something to "install" even if it's standalone?
> i tried downloading a game (also only one binary) once but i couldn't start it at all.

Just put it in `/usr/local/bin`, after removing the previous `yt-dlp`.
Rename to `yt-dlp` and use it as you did before.

> > No, that .exe has Python 3.8 built in.
> 
> so i assume, the same will be the case once 3.8 support dropping (the last version for windows 7)? in other words, will i still be able to use yt-dlp standalone on windows 7 after 2024 (python 3.8 eol)?

Eventually, 3.8 will not be supported by yt-dlp any more.
See Pukkandan's [comment](https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1162980553) here.

**Comment by Vangelis66** [comment]
> will the standalone exe for windows 7 then have python 3.9 or 3.10 or whatever newer version have built in? 
in short: will there still be **a working standalone exe for windows 7 after python 3.8 is eol**?

Well, I don't mean to be rude 😄 , but an easy G-search would have told you that **the last CPython version compatible with Windows 7 SP1 is the 3.8.x branch**; IOW, 3.9.x+ won't launch in said OS (however, there's a "hack" hosted in GitHub somewhere that enables 3.9.x binaries to launch under Win7SP1...).

The last ever released `yt-dlp.exe` binary (x64) with embedded 3.8.x will continue to "work" in Win7SP1, however newer bug fixes, extractors, etc. won't be available anymore for the OS... 
FWIW, Win7 SP1. despite its popularity, will **move out of ESU next January**, so major vendors will drop support for it for good (among them, Google...).

Kind regards.

**Comment by pukkandan** [comment]
btw, tangentially related; I have been thinking of bumping `yt-dlp.exe`'s Python version to `3.10`. This will bring all the latest performance improvements (not sure how much - will need to test) and bugfixes/security updates to **most** windows users, but make it not work in Windows 7. However win7 users can still use `yt-dlp_x86.exe` which will continue to be built on the minimum supported version (currently 3.7). Opinions? Does it matter to the win7 users here that the binary is x86?

**Comment by honx4amiga** [comment]
> However win7 users can still use `yt-dlp_x86.exe` which will continue to be built on the minimum supported version (currently 3.7). Opinions? Does it matter to the win7 users here that the binary is x86?

is _x86.exe technically the same (except for being 32bit)? is usage the same, are parameters the same? i made four batch files i use very often:
mp3.bat:
yt-dlp --extract-audio --audio-format mp3 --no-mtime --output "%%(title)s.%%(ext)s" %1
mp4.bat:
yt-dlp --format "bestvideo[ext=mp4]+bestaudio[ext=m4a]" --no-mtime --output "%%(title)s.%%(ext)s" %1
mp3_playlist.bat:
yt-dlp --extract-audio --audio-format mp3 --no-mtime --output "%%(title)s.%%(ext)s" --download-archive id.txt %1
mp4_playlist.bat:
yt-dlp --format "bestvideo[ext=mp4]+bestaudio[ext=m4a]" --no-mtime --output "%%(title)s.%%(ext)s" --download-archive id.txt %1

will these continue to work with _x86 (32bit binary) if i replace yt-dlp by yt-dlp_x86 (or rename yt-dlp_x86 )?

and in linux mint 19 (tara) xfce: how do i uninstall/remove now obsolete yt-dlp?

**Comment by Vangelis66** [comment]
> I have been thinking of bumping `yt-dlp.exe`'s Python version to `3.10`.

I bet this will take with it `yt-dlp_min.exe` and `yt-dlp_win.zip`, wouldn't it?

> PS: 3.11 release [is just around the corner](https://peps.python.org/pep-0664/)

Expected:
    **3.11.0 final**: Monday, 2022-10-03

> However win7 users can still use `yt-dlp_x86.exe`, which will continue to be built on the minimum supported version (currently 3.7). 
Opinions? Does it matter to the win7 users here that the binary is x86?

If you delegate the **still plentiful** Win7 users to py3.7 builds, then they'd fall at a worse perspective **support-wise**, compared to the status quo of being relied upon py3.8 builds; as already posted in this thread, PSF will EOL 3.8 in Oct 20**24**, but 3.7 is EOL'd in July 20**23**...

As far as architecture is concerned, most **Win7** users should be already at **x64**, which can run 32-bit apps fine,; a small percentage of them should be on **32-bit**, especially those on older H/W (upgraded XP systems come to mind... 😉 ); but one knows the memory limitations of 32-bit apps on a 64-bit OS...

My second laptop still runs **Win7 x64**, so I might get affected by this newer "consideration', TBH... 😒

**Comment by pukkandan** [comment]
> is _x86.exe technically the same (except for being 32bit)? is usage the same, are parameters the same? i made four batch files i use very often:

Yes, the x86 build is the exact same as `yt-dlp.exe` except that it is built on a 32bit architecture

> I bet this will take with it yt-dlp_min.exe and yt-dlp_win.zip, wouldn't it?

I would prefer to update them as well, but wouldn't mind keeping these in 3.8 if there is **real use-case** for doing so

> If you delegate the still plentiful Win7 users to py3.7 builds, then they'd fall at a worse perspective support-wise, compared to the status quo of being relied upon py3.8 builds; as already posted in this thread, PSF will EOL 3.8 in Oct 2024, but 3.7 is EOL'd in July 2023...

No, since when we drop 3.7 support, the x86 build will obviously be moved to 3.8!

> but one knows the memory limitations of 32-bit apps on a 64-bit OS...

That it can use "only" 2GB of RAM? Do you really use yt-dlp in a way that is affected by this?

**Comment by gamer191** [comment]
> The build has to be created against the minimum version we want to support. Since the [oldest Linux GHA supports is Ubuntu 18.04](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) which [is based on Debian 10](https://askubuntu.com/a/445496), it won't be possible for us to create standalone builds for Debian 9

Doesn't that mean that, if we drop Python 3.6 support, then as soon as Github removes support for Ubuntu 18.04 (which will likely be quite soon because Ubuntu 22.04 support is [already in beta](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources)), our Linux binaries will stop supporting a large number of Linux distros?

> and how do i uninstall/remove previous yt-dlp completely in linux mint 19 (tara) xfce?

@honx4amiga can't you just delete it from `/usr/local/bin`?

> btw, tangentially related; I have been thinking of bumping `yt-dlp.exe`'s Python version to `3.10`. This will bring all the latest performance improvements (not sure how much - will need to test) and bugfixes/security updates to **most** windows users, but make it not work in Windows 7. However win7 users can still use `yt-dlp_x86.exe` which will continue to be built on the minimum supported version (currently 3.7). Opinions? Does it matter to the win7 users here that the binary is x86?

IMO we should solve that problem by having a 64-bit python-latest binary and a 64-bit python 3.8 binary. But the release-files section of the readme is getting quite long, so we'll need to first re-write it, perhaps using `<summary>` and `<details>` tags (I'm going to create a PR to do that, but probably not for a while, especially because I'm going away tomorrow for a week).


> > If you delegate the still plentiful Win7 users to py3.7 builds, then they'd fall at a worse perspective support-wise, compared to the status quo of being relied upon py3.8 builds; as already posted in this thread, PSF will EOL 3.8 in Oct 2024, but 3.7 is EOL'd in July 2023...
> 
> No, since when we drop 3.7 support, the x86 build will obviously be moved to 3.8!


But afaik there hasn't been any decision (yet) to drop Python 3.7 support as soon as the PSF does


> > but one knows the memory limitations of 32-bit apps on a 64-bit OS...
> 
> That it can use "only" 2GB of RAM? Do you really use yt-dlp in a way that is affected by this?

I'm no expert, and I haven't tested this, but wouldn't that slow down pycryptodomex downloads?

**Comment by pukkandan** [comment]
> Doesn't that mean that, if we drop Python 3.6 support, then as soon as Github removes support for Ubuntu 18.04 (which will likely be quite soon because Ubuntu 22.04 support is [already in beta](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources)), our Linux binaries will stop supporting a large number of Linux distros?

https://github.com/pyinstaller/pyinstaller/discussions/5669 seems to have a couple solutions for this. I'll look into it

**Comment by honx4amiga** [comment]
> @honx4amiga can't you just delete it from `/usr/local/bin`?

since i don't know which files were copied/installed to where this is no solution. i doubt if yt-dlp install came with just one file...
so there must be a proper way to uninstall yt-dlp which removes ALL files and folders created during yt-dlp install completely...

**Comment by pukkandan** [comment]
> > @honx4amiga can't you just delete it from `/usr/local/bin`?
> 
> since i don't know which files were copied/installed to where this is no solution. i doubt if yt-dlp install came with just one file... so there must be a proper way to uninstall yt-dlp which removes ALL files and folders created during yt-dlp install completely...



yt-dlp is a portable program. You "install" it by either copying the binary to PATH or by using any package manager. If you installed with a PM, use the same to uninstall. If you simply downloaded the binary to PATH, just delete it.

**Comment by honx4amiga** [comment]
> yt-dlp is a portable program. You "install" it by either copying the binary to PATH or by using any package manager. If you installed with a PM, use the same to uninstall. If you simply downloaded the binary to PATH, just delete it.

i'm talking about the currently installed now obsolete version based on python 3.6, which is definitely NOT just one executable. so this version i am talking about is definitely NOT "portable"! as it's very long ago i installed yt-dlp i can't remember what i had to do to install it (can't remember which install instruction i read/followed). but it was definitely NOT a "simple download". i only remember that everything had to be done in terminal window, there was no package manager involved. but i can't remember if i used "sudo apt-get install yt-dlp" or something else to get this thingamajigs installed.

btw. "simple downloads" of executables in linux NEVER worked for me before. after downloading any executables i never was able to actually execute these "executables". simply nothing happend when trying to launch a downloaded executable...

**Comment by pukkandan** [comment]
`apt` IS a package manager 🤦. This is getting off topic. If you don't know how you installed your yt-dlp, there is nothing we can do to help. You are better off asking in some Linux forums

**Comment by fstirlitz** [comment]
> actually, apparently amd64 and aarch64 are the same thing

ಠ_ಠ

**Comment by honx4amiga** [comment]
> `apt` IS a package manager facepalm. This is getting off topic. If you don't know how you installed your yt-dlp, there is nothing we can do to help. You are better off asking in some Linux forums
i think this was how i got yt-dlp installed back in the day:
sudo wget https://github.com/yt-dlp/yt-dlp/releases/latest/download/yt-dlp -O /usr/local/bin/yt-dlp
sudo chmod a+rx /usr/local/bin/yt-dlp

is there only that yt-dlp in /usr/local/bin to be deleted or replaced by that standalone version? nothing else?

**Comment by pukkandan** [comment]
> > Doesn't that mean that, if we drop Python 3.6 support, then as soon as Github removes support for Ubuntu 18.04 (which will likely be quite soon because Ubuntu 22.04 support is [already in beta](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources)), our Linux binaries will stop supporting a large number of Linux distros?
> 
> [pyinstaller/pyinstaller#5669](https://github.com/orgs/pyinstaller/discussions/5669) seems to have a couple solutions for this. I'll look into it

https://github.com/pukkandan/yt-dlp-dev/commit/71086cfaa24e8650543eb3a92cdaeecb38bf9458 will allow us to build under ubuntu18 even after GHA drops it - as long as [deadsnakes/ppa](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa) supports it. It is even possible for us to continue building it after they drop support if someone is willing to make a PR for automating [these instructions](https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1163743264) (or their ubuntu equivalent)

Test build: https://github.com/pukkandan/yt-dlp-dev/releases/download/2022.06.26.111239/yt-dlp_linux

**Comment by gamer191** [comment]
> It is even possible for us to continue building it after they drop support if someone is willing to make a PR for automating [these instructions](https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1163743264) (or their ubuntu equivalent)

Theoretically I imagine that that would also add support for Debian 9 and other legacy linux distros (assuming that the docker container is equivalent to Debian 9 or older)

**Comment by nicolaasjan** [comment]
> [pukkandan@71086cf](https://github.com/pukkandan/yt-dlp-dev/commit/71086cfaa24e8650543eb3a92cdaeecb38bf9458) will allow us to build under ubuntu18 even after GHA drops it - as long as [deadsnakes/ppa](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa) supports it. It is even possible for us to continue building it after they drop support if someone is willing to make a PR for automating [these instructions](https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1163743264) (or their ubuntu equivalent)
> 
> Test build: https://github.com/pukkandan/yt-dlp-dev/releases/download/2022.06.26.111239/yt-dlp_linux

At least in Debian 10 (ships with Python 3.7.3, so still supported), it works:
(I dont have a Debian 9 VM to test)

```python
yt-dlp -vU
[debug] Command-line config: ['-vU']
[debug] User config "/home/nico/.config/yt-dlp/config": ['-o', '/dev/shm/test-dlp/%(title)s.%(ext)s', '-f', 'bestvideo[height<=1080][ext=mp4][vcodec!*=av01]+bestaudio[ext=m4a]/best[ext=mp4]/best', '--no-mtime', '--add-metadata', '--convert-thumbnails', 'jpg', '--embed-thumbnail', '--ppa', 'ThumbnailsConvertor:-q:v 1', '--ppa', 'ffmpeg:-metadata synopsis=""', '--force-ipv4', '--sponsorblock-remove', 'all', '--console-title', '--downloader', 'aria2c', '--downloader-args', 'aria2c:-x 10 -s 10 -j 10 -k 1M --log-level=info --file-allocation=none']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.06.26.111239 (linux_exe)
[debug] Python version 3.10.5 (CPython 64bit) - Linux-4.19.0-20-amd64-x86_64-with-glibc2.28
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg N-107213-gfed07efcde-20220622 (setts), ffprobe N-107213-gfed07efcde-20220622
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, secretstorage-3.3.2, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.06.22.1, Current version: 2022.06.26.111239
yt-dlp is up to date (2022.06.26.111239)
```

**Comment by gamer191** [comment]
> (I don't have a Debian 9 VM to test)

Pukkandan's test build is very unlikely to work on Debian 9, because it was compiled on Ubuntu 18.04. If someone wanted to create a PR to add support for Debian 9, they could probably use [this solution](https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1166522227). Otherwise, they can manually follow [these instructions](https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1163743264) to install it on their personal computer.

**Comment by nicolaasjan** [comment]
> Pukkandan's test build is very unlikely to work on Debian 9, because it was compiled on Ubuntu 18.04.

Indeed. I see now that the [deadsnakes/ppa](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa) has no versions for Ubuntu 16 and lower.

**Comment by csehszlovakze** [comment]
Asking for advice: how can I tell yt-dlp **only** to use python 3.8 specifically? See, if I change /usr/bin/python3 to point to 3.8, then a lot of things break under Kubuntu 18.04, which still depends on python 3.6. I do plan on upgrading to 22.04 eventually, but what should I do until then?

**Comment by nyuszika7h** [comment]
Either change the shebang to `#!/usr/bin/env python3.8` (has to be redone after every update) or do `alias yt-dlp='python3.8 -m yt_dlp'`. (Or you could install it in a separate virtualenv with pyenv/pipx.)

**Comment by csehszlovakze** [comment]
I tried these, both the alias (sans -m) and the shebang change work when using yt-dlp alone, but if I'm using it in tandem with mpv and SMPlayer, then only the overriding of /usr/bin/python3 works, **but that breaks apt completely.**

Should I just switch to the non-python binary on my current system and resume using this one once I'm on 22.04.1?

**Comment by honx4amiga** [comment]
if there so many problems occur in the context of dropping python 3.6 support (see csehszlovakze comments) why is switching to 3.7 necessary? is there any feature depending on python 3.7? wouldn't it be easier to just keep maintaining 3.6 support since there are so many distros still using python 3.6 (not just linux mint 19 i use)?

**Comment by thinkpad4** [comment]
So after reading all this I've come to the understanding that you will be keeping the program on 3.8 until at least 2024 and will be keeping a x64 and x86 version of the program both on 3.8 until at least 2024.  Right?

**Comment by pukkandan** [comment]
@csehszlovakze You can use the standalove executable. Or, you could install with pip. Then yt-dlp will always run only on the Python version you installed it on - `python3.8 -m pip install yt-dlp`.

@honx4amiga The reasons have been explained in the OP and subsequent comments. And what "so many problems?". I believe almost all concerns in this thread have already been addressed. Your questions about "uninstalling yt-dlp" were left unanswered because [they are off topic]()! The only real unsolved issue afair is https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1163743264 and [PRs are welcome to solve it](https://github.com/yt-dlp/yt-dlp/issues/3764#issuecomment-1166499925)

@thinkpad4 We are not discussing anything about Python 3.8. It becomes EOL on 2024, but we may decide to drop it early or to keep supporting it for longer depending on circumstances

**Comment by FranciscoPombal** [comment]
As usual, a widely-used piece of free software gets harassed by GitHub's resident vocal minority of egotistical neo-luddites, lest it cease to support their dangerously outdated setups that they invest 10x less energy into updating than what they do posting in these threads. I have seen some of the very same posters ITT doing the exact same thing in other projects.

As of even a few years ago, there is NO legitimate reason to purposefully stick to using Vista/7 or older as a consumer (ofc some PoS and ATMs still run those, but that has nothing to do with yt-dlp) other than as a retrocomputing hobby - in which case there is no right to complain about dropping EOL versions of dependencies here - fork the project and do with that as you wish.

Hardware access is a non-issue - you can install a modern Linux distro on shitboxes from 2006 or even earlier (I still have a 2 core Atom from 2006 or so + 2 GiB RAM, and it's running Ubuntu 22.04). Windows 10 can also be installed on quite old hardware (but obviously not quite as old with a reasonable computing experience). And you can do everything you want on those systems.
There is no excuse. Stop holding other users hostage and deprived of improvements from newer versions of software by pressuring project developers to bend over backwards with unreasonable backwards compatibility requests because of your laziness.

Also, lmao @ people claiming 16.04 is still "supported", and other equivalent claims - it's not, at least not in the sense that matters here. ESM is PAID support. If you really need yt-dlp on such systems, PAY Canonical or someone else to maintain it for you, or maintain it yourself, don't impose your PAID requirements on the rest of the userbase of a FOSS project.

As always, my sincerest thank you to the maintainers for putting up with this bullshit in addition to all their efforts all around.

/rant over

**Comment by honx4amiga** [comment]
> As of even a few years ago, there is NO legitimate reason to purposefully stick to using Vista/7

what about refusing m$ spying tactics? windows 10 is fckng SPYWARE and there are FORCED "updates" you can't avoid installing since they are FORCED, bringing even more SPYWARE into your system and even being able to brick your system if a FORCED "update" goes bad.
windows 11 is even WORSE, requires a FORCED m$ account for even being able to install that bs so m$ can sniff even more data from its users. you pay money for 10 and 11 and in addition to paying money you also have to pay your personal data!
that excessive spying on users started with windows 8.1 and in addition 8.1 is also ugly.
in other words: windows 7 is the LAST good windows based operating system. 8.1, 10 and 11 are crap and far away from "user-friendly"! this is my LEGITIMATE reason to stick with windows 7!

**Comment by nyuszika7h** [comment]
Just use Windows 10 LTSC or Linux, you'll be far better off.

**Comment by honx4amiga** [comment]
> Just use Windows 10 LTSC

ltsc is for companies only (enterprise edition), NOT available for private users (home/pro), and there are major disadvantages in ltsc...

**Comment by gamer191** [comment]
> As usual, a widely-used piece of free software gets harassed by GitHub's resident vocal minority of egotistical neo-luddites, lest it cease to support their dangerously outdated setups that they invest 10x less energy into updating than what they do posting in these threads. I have seen some of the very same posters ITT doing the exact same thing in other projects.

Mate, do you have any examples of messages in this thread that would be considered harassment? After all, the original message in this thread included, at the top of it:
> This is not an announcement. We are simply discussing the possibility for now

meaning that Pukkandan, the maintainer of this project, wanted opinions about whether he should remove support for Python 3.6.

> there is NO legitimate reason to purposefully stick to using Vista/7 or older as a consumer

Personally I use Windows 7 because it still has [Windows Media Center](https://en.wikipedia.org/wiki/Windows_Media_Center), but that's probably just me 🤣. That said, I use yt-dlp on a different computer, so I don't really give a shit if yt-dlp removes Windows 7 support

> Ubuntu ESM is PAID support

I'm pretty sure that it's free for non-commercial purposes, but don't let the facts get in the way of a good argument



> this is my LEGITIMATE reason to stick with windows 7!

@honx4amiga I thought you said before that you use Linux. If you use both, out of interest, is there any specific reason why you keep Windows 7? I'm just curious

**Comment by pukkandan** [comment]
> > Ubuntu ESM is PAID support
> 
> I'm pretty sure that it's free for non-commercial purposes, but I could be wrong

They mean that the extended **support** is paid. Not that you need to pay for the software itself

**Comment by dirkf** [comment]
https://ubuntu.com/blog/ubuntu-16-04-lts-transitions-to-extended-security-maintenance-esm:
>ESM is also available to personal users on up to three machines and Ubuntu members on 50 machines.

**Comment by csehszlovakze** [comment]
quote @pukkandan 
>You can use the standalove executable. Or, you could install with pip. Then yt-dlp will always run only on the Python version you installed it on - `python3.8 -m pip install yt-dlp`.

Thanks for the suggestion, the standalone executable seems to be a lot better than the default release! I see fewer errors when streaming with SMPlayer/mpv now. I do wish it included manpages and autocomplete and updated those also when a new version comes out, but that's just me.

.

quote @honx4amiga 
>i use both. main reasons for windows 7 are ms office for exchange with job (on linux mint i only have libre office available) and of course games. third reason is streaming (youtube or similar). on linux mint i can't stream full hd (1080p) without stuttering. only on windows i can stream full hd movies smooth without any problems.

* Try streaming videos with mpv, not the browser or vlc, from my anecdotal testing that provides the best performance.
* Have you tried the web/electron version of msoffice yet? That one will run on linux, but I've no idea what it's like considering I switched to LibreOffice back when msoffice introduced their ribbon interface... (yeah that was ages ago)

**Comment by joeschmoe40** [comment]
2 questions:
1) I see references above to "the standalone executable".  Is that Windows-only or does it also work on (i.e., is available for) Linux (Debian/Ubuntu) ?
2) Now that we've all accepted (and adjusted to) the move to 3.7 (minimum), what is the timetable/schedule for moving to 3.8 (and etc) ?  I assume it will happen, just curious as to when.

**Comment by pukkandan** [comment]
> * I see references above to "the standalone executable".  Is that Windows-only or does it also work on (i.e., is available for) Linux (Debian/Ubuntu) ?

There are executables for Windows, Linux and Mac. See https://github.com/yt-dlp/yt-dlp#release-files

> * Now that we've all accepted (and adjusted to) the move to 3.7 (minimum), what is the timetable/schedule for moving to 3.8 (and etc) ?  I assume it will happen, just curious as to when.

There is no plans. We'll think about it when the need arises

**Comment by joeschmoe40** [comment]
Quote:
There are executables for Windows, Linux and Mac. See https://github.com/yt-dlp/yt-dlp#release-files
EndQuote:
I am looking for a "standalone executable" (of yt-dlp) for Debian/Ubuntu systems.
I clicked on that link, but couldn't find anything to download.  Please help.
Note: I am very good with computers and scripting (just in case you were assuming otherwise), but not so good with Githubbing.  I usually get lost there.
    On Sunday, July 31, 2022, 09:03:43 AM EDT, pukkandan ***@***.***> wrote:  
 
 


   
   - I see references above to "the standalone executable". Is that Windows-only or does it also work on (i.e., is available for) Linux (Debian/Ubuntu) ?


There are executables for Windows, Linux and Mac. See https://github.com/yt-dlp/yt-dlp#release-files

   
   - Now that we've all accepted (and adjusted to) the move to 3.7 (minimum), what is the timetable/schedule for moving to 3.8 (and etc) ? I assume it will happen, just curious as to when.


There is no plans. We'll think about it when the need arises

**Comment by pukkandan** [comment]
![image](https://user-images.githubusercontent.com/6623388/182034402-a31b2891-62cd-43b7-90cc-65aa3a6670c9.png)

**Comment by gamer191** [comment]
@joeschmoe40 just use this direct link: https://github.com/yt-dlp/yt-dlp/releases/latest/download/yt-dlp_linux

(and you can update it using `yt-dlp --update`)

**Comment by joeschmoe40** [comment]
Also, is there a way for me to build my own "standalone" executable, say for a Linux platform other than x64 (e.g., ARM) ?

**Comment by pukkandan** [comment]
> When I clicked on "Installation" at the top of it, nothing happened. 

Issue with browser? All the links work for me

> Just out of curiosity, which version of Python is bundled up inside that very large (Linux, x64)

Currently 3.10

> Is there a yt-dlp option/command to make it tell you which version of Python it thinks it is running under?

`yt-dlp -v`

> Also, is there a way for me to build my own "standalone" executable, say for a Linux platform other than x64 (e.g., ARM) ?

See https://github.com/yt-dlp/yt-dlp#standalone-pyinstaller-builds

PS: The docs say "Windows/MacOS builds". This is wrong. I forgot to update the docs when adding Linux build

**Comment by joeschmoe40** [comment]
OK - all good now.

**Comment by joeschmoe40** [comment]
Actually, there is another problem - with the Linux "standalone executable".  Note: This is somewhat "off topic" to this thread; I don't know exactly what the topicality rules are here.

Anyway, I've found that while the Linux "standalone executable" does indeed "self-update", the resulting new file is not marked executable (in the Linux/Unix sense).  That is, its "mode" is 644 when it should be 755.  This problem does not manifest with the "pure Python" (Actually, zipped) version; there, the resulting file will be executable (assuming the original was).

I believe this is a result of the fact that replacing a running (binary) executable is tricky, on all platforms, and that in order to do it, some trickery involving downloading the new file to another name and then renaming it back to the original is involved.  And that somewhere along the way, the "x" bits on the file get lost.  So, the result is you have to do an explicit "chmod +x yt-dlp_linux" after using "-U" to update it.

It would be good if this could be fixed.

**Comment by pukkandan** [comment]
@joeschmoe40 Could you pls open a new issue for this

**Comment by codenamedpkt** [comment]
Right, drop support for outdated versions of Windows which no full offical support by Microsoft there right now. (Vista, 7, 8, 8.1)

**Comment by honx4amiga** [comment]
> Right, drop support for outdated versions of Windows which no full offical support by Microsoft there right now. (Vista, 7, 8, 8.1)

NOOOOO!!!!! there are still people (like me) using windows 7!

**Comment by Vangelis66** [comment]
Me humbly thinks this "**Discussion**" turned "**Announcement**" has served its purpose/ran its course; all the **significant decisions have been made by the devs and already put in effect** (**Py3.6 support** has been removed); I propose it be locked, as it, apparently, now only attracts spam/trolls 😞 ...

FTR, **Win8.1** is still under **Extended Support** until next January, though even that is a moot point, because **Py3.10** still supports Win8+; **3.11-stable**, to be released **before the start of 2023**, will - most likely - retain that support; I don't expect **Win8.1** to be dropped by **PSF** sooner than **3.12-final** is released (on 2023-10-02).

**Comment by MrRawes** [comment]
https://github.com/yt-dlp/yt-dlp/issues/3765 says this issue is "Active", shouldn't this issue be in the "Closed" section

**Comment by pukkandan** [comment]
In case anyone's still using 3.6, I've made [one more release](https://github.com/yt-dlp/yt-dlp/releases/tag/2022.08.18.36) because of https://github.com/yt-dlp/yt-dlp/issues/4635. Running `yt-dlp -U` on 3.6 will automagically update to it

**Comment by bubbleguuum** [comment]
If you need to restore Python 3.6 support for some reason, you can do it reverting these 2 commits:

```
git revert 6929b41a216e20f0498cbd99880b17eab16777c9
git revert 6a7d3a0a0981d05903e70bcb31fc3f9438eedf22
```

**Comment by kabeza** [comment]
Hi
I'm running on Ubuntu 20.04 and have installed
python 2.7
python 3.6
python 3.9
I've installed Tartube (works fine with yt-dlp) but I get an error regarding python 3.6
**DeprecationWarning: Support for Python version 3.6 has been deprecated**

Is there anyway to make it work by default with the python 3.9 version I have?

I don't want to set it as default as it may break my Ubuntu install, etc. (as I've read somewhere)

Thanks

**Comment by joeschmoe40** [comment]
{Quote}
I'm running on Ubuntu 20.04 and have installed
python 2.7
python 3.6
python 3.9
I've installed Tartube (works fine with yt-dlp) but I get an error regarding python 3.6
DeprecationWarning: Support for Python version 3.6 has been deprecated

Is there anyway to make it work by default with the python 3.9 version I have?

I don't want to set it as default as it may break my Ubuntu install, etc. (as I've read somewhere)
(End Quote}

Note: I don't know how to do quoting in this editor, so I did the best I could above.

Anyway, I had this same problem with an Ubuntu machine - where the default (In fact, only) version of Python installed was (is) 3.6 and, as we all know by now, that's not good enough.  Like you, I don't want to mess around with installing other versions, because I don't want to risk breaking my system.

I solved my issue by downloading and running the "stand alone executable" version.  This seems to work fine, and finesses all the versioning issues.  I suggest you do the same.  Note that this works fine as long as you are running on an X64 system, which is most people, most of the time.  It is not a fully good solution for me, since many of my Linux machines are ARM, but it is good enough for the one Ubuntu system that I have.

**Comment by Vangelis66** [comment]
@joeschmoe40 wrote: 

> Note: I don't know how to do quoting in this editor, so I did the best I could above.

![quote1](https://user-images.githubusercontent.com/9669492/191642109-b25e2625-1da1-454e-a145-1d16f677a60e.JPG)

Alternatively: 

https://commonmark.org/help/tutorial/05-blockquotes.html

![quote2](https://user-images.githubusercontent.com/9669492/191642591-8c56a97d-9b31-4af2-94f6-bbc408c25c61.JPG)

**Comment by pukkandan** [comment]
Those who were unable to use `yt-dlp_linux` due to libc version, check out https://github.com/yt-dlp/yt-dlp/pull/5423

**Comment by FlaminHat** [comment]
How do I upgrade Python like it keeps telling me to do???

**Comment by FlaminHat** [comment]
Sill says I'm on Python 3.6
